### ISSUE TITLE
Can't find ConnectionProvider (Wallet No Longer Works)

### ISSUE THREAD

**Issue by tiagoappetite** [solana-labs/wallet-adapter#532]
Stopped working for me this night in all projects I've after update (React's Next.js and tried on Svelte too)

**Error [ERR_MODULE_NOT_FOUND]: Cannot find module /app/node_modules/@solana/wallet-adapter-react/lib/esm/ConnectionProvider**

**Comment by danvernon** [comment]
+1

**Comment by josip-volarevic** [comment]
Just spent a while trying to troubleshoot the same problem. Took me a while to figure out that there was a new release yesterday that (most likely?) broke stuff.

I'm getting the following error when trying to run my nextjs app:
```
error - Error [ERR_MODULE_NOT_FOUND]: Cannot find module 'C:\Users\username\project-name\node_modules\@solana\wallet-adapter-material-ui\lib\esm\useWalletDialog' imported from C:\Users\username\project-name\node_modules\@solana\wallet-adapter-material-ui\lib\esm\index.js
    at new NodeError (node:internal/errors:387:5)
    at finalizeResolution (node:internal/modules/esm/resolve:429:11)
    at moduleResolve (node:internal/modules/esm/resolve:1006:10)
    at defaultResolve (node:internal/modules/esm/resolve:1214:11)
    at nextResolve (node:internal/modules/esm/loader:165:28)
    at ESMLoader.resolve (node:internal/modules/esm/loader:844:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:431:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:76:40)
    at link (node:internal/modules/esm/module_job:75:36) {
  code: 'ERR_MODULE_NOT_FOUND',
  page: '/'
}
```

My repository can be found here: https://github.com/josip-volarevic/narentines-holders-only/tree/wallet-adapter-issue
I'll leave the `wallet-adapter-issue` branch live for reproduction purposes until the issue is fixed

**Comment by tiagoappetite** [comment]
I’ve been working yesterday on both Next.js and SvelteKit (this one uses Vite instead Webpack) apps and it was working fine. 

This morning after updates / trying new fresh new setups none works. Tried with different packet managers too (yarn and pnpm)

**Comment by josip-volarevic** [comment]
@tiagoappetite and your repos work with older package versions?

**Comment by anoushk1234** [comment]
Same issue

**Comment by grhbit** [comment]
Same issue.

And these are the last versions working for me.
```JSON
{
    "@solana/wallet-adapter-base": "0.9.13",
    "@solana/wallet-adapter-react": "0.15.15",
    "@solana/wallet-adapter-react-ui": "0.9.13",
    "@solana/wallet-adapter-wallets": "0.17.2"
}

**Comment by danvernon** [comment]
> @tiagoappetite I assume your repos work with older package versions?

i think some dependencies are causing it, if i yarn in any project that did works - it breaks

**Comment by jordansexton** [comment]
Thanks for all the reports, I'll work on this today.

First I need to isolate the issue, to see if it's reproducible within the repo, and if not, what causes it.

Can you test something and tell me if it works? (I'm typing shell cmds from my phone apologies for any mistakes) --

```
git clone https://github.com/solana-labs/wallet-adapter.git
cp -R wallet-adapter/packages/starter/nextjs-starter .
cd nextjs-starter
npm install
npm run dev
```

**Comment by aslaker** [comment]
@jordansexton I just rant those command and here is the error that I get.  I wonder if it has to do with some server side specific code in a dependency trying to run on the frontend side in NextJS.  I noticed a similar issue yesterday with @gokiprotocol/walletkit.  Any chance it is coming from a dependency that both projects rely on?

`Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/Users/adamslaker/personal/vanth/test/nextjs-starter/node_modules/@solana/wallet-adapter-base/lib/esm/adapter' imported from /Users/adamslaker/personal/vanth/test/nextjs-starter/node_modules/@solana/wallet-adapter-base/lib/esm/index.js`

**Comment by tiagoappetite** [comment]
> 

Following those steps I get similar error:

`Error: Cannot find module '/Users/tiago/wallet-adapter/packages/starter/nextjs-starter/node_modules/@solana/wallet-adapter-base/lib/esm/adapter' imported from /Users/tiago/wallet-adapter/packages/starter/nextjs-starter/node_modules/@solana/wallet-adapter-base/lib/esm/index.js`

-

And in case this helps, here's the issues getting on non-react frameworks that were working fine yesterday.

**Nuxt (Vue)**
`Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/Users/tiago/nuxt3-solana/node_modules/@solana/wallet-adapter-avana/lib/esm/adapter' imported from /Users/tiago/nuxt3-solana/node_modules/@solana/wallet-adapter-avana/lib/esm/index.js`

**Svelte**
`This file is being treated as an ES module because it has a '.js' file extension and '/Users/tiago/my-app/node_modules/@solana/wallet-adapter-base/package.json' contains "type": "module". To treat it as a CommonJS script, rename it to use the '.cjs' file extension.`

**Comment by kgilliam125** [comment]
Curious thing I'm seeing (and seen in other comments), reverting to a prior version of wallet-adapter does not seem to resolve this issue. This makes me think that the problem is with a dependency which is set via `^x.y.z` and would grab the most recent version per the rule.

Look through the `react` and `base` [package.json](https://github.com/solana-labs/wallet-adapter/blob/78b42f46baa200edae60d6a48defa7cd5fd77541/packages/core/react/package.json#L30) the only one that stands out it `@solana/web3.js` which released a new version, [1.53.1](https://www.npmjs.com/package/@solana/web3.js), a bit before we started seeing wallet-adapter build failures. 

Seems unlikely that the web3.js build itself is bad, but could possibly be a small change in usage that is cascading?

**Comment by jordansexton** [comment]
What if you use npm overrides / yarn resolutions to force a specific pinned exact version?

**Comment by jordansexton** [comment]
I'll be on my computer soon to fix this. Offhand, it sounds like what's going on is that this is related to an aspect of the change I made yesterday.

We still ship CJS and ESM builds, but the ESM build outputs are no longer .mjs files, just .js files. Some build tools might have been broken by this change.

**Comment by kgilliam125** [comment]
> What if you use npm overrides / yarn resolutions to force a specific pinned exact version?

Put this in real quick and ran a build, still the same issue as above. May have been a false lead. Will dig more on my end when I get back to my computer.

**Comment by tiagoappetite** [comment]
The problem seems to be from the `@solana/wallet-adapter-base`

I've a build from 5 days ago with version `0.9.12` it works.
Doing an update changes it to version `0.9.14`

That said even if you bring it down to `0.9.12` it still don't works as all the wallets have version `0.9.14` as dependencies

**Comment by jordansexton** [comment]
Okay, an informational update --

This fails:
```
git clone https://github.com/solana-labs/wallet-adapter.git
cp -R wallet-adapter/packages/starter/nextjs-starter .
cd nextjs-starter
npm install
npm run dev
```

This works:
```
git clone https://github.com/solana-labs/wallet-adapter.git
cp -R wallet-adapter/packages/starter/react-ui-starter .
cd react-ui-starter
npm install
npm start
```

The difference here is that Next.js vs Parcel. My guess is that Parcel is handling imports without explicit .mjs extensions, and Next isn't.

Working on configuration that will resolve this.

**Comment by jordansexton** [comment]
Another data point -- the `create-react-app-starter` package works. But because Webpack 5 _does not_ handle imports of ESM without .mjs extensions, we have to configure it to do so: https://github.com/solana-labs/wallet-adapter/blob/78b42f46baa200edae60d6a48defa7cd5fd77541/packages/starter/create-react-app-starter/config-overrides.js#L15-L20

Next.js _used to use_ Webpack 5, but I'm confused as to whether it does anymore anymore.

Their docs indicate it's using their Rust compiler, SWC: https://nextjs.org/docs/advanced-features/compiler

However, their docs _also_ indicate you can configure Webpack: https://nextjs.org/docs/api-reference/next.config.js/custom-webpack-config

This is what I'll try next.

**Comment by jordansexton** [comment]
Okay, so it appears the difference is that SWC is a replacement for Babel, but Webpack 5 is still used.

Adding this `next.config.js` file:

```js
/** @type {import('next').NextConfig} */
module.exports = {
    reactStrictMode: true,
    webpack(config, context) {
        console.log(config.module.rules);
        process.exit(1);

        return config;
    },
};
```

We get this:
```js
[
  { test: /\.m?js/, resolve: { fullySpecified: false } },
  /* ... */
]
```

This indicates Next is already configured by default to handle .js and .mjs files imported without an extension. This is good and what we'd expect, but unfortunately suggests the problem lies elsewhere.

**Comment by jordansexton** [comment]
Curiously, this issue is also affecting `@solana/spl-token`, which _does_ build .mjs files:
https://github.com/solana-labs/solana-program-library/issues/3502

**Comment by tiagoappetite** [comment]
I've just tried start a new React app with Vite... and it also worked just fine!!

Which puzzles me given that Nuxt.js uses Vite and was also giving an error similar to Next.

**Comment by jordansexton** [comment]
It seems like this _may_ be related to an open issue with Next.js: https://github.com/vercel/next.js/issues/39375

I also tried using the TS config option, `"moduleResolution": "nodeNext"` and it seems to have no effect. Related: https://github.com/vercel/next.js/issues/38854

_However_ -- I was able to fix the issue locally in a project by manually changing all the imports without extensions to `.js`.

So what I might need to do is change all of the imports in Wallet Adapter to do this. I'll then have to make sure this doesn't break Parcel, Webpack, etc.

**Comment by tiagoappetite** [comment]
I was actually trying something similar.

Those are the default TS config settings on a new React app (Vite)... tried replicate them over Next / Nuxt without any success.

```
  "compilerOptions": {
    "target": "ESNext",
    "useDefineForClassFields": true,
    "lib": ["DOM", "DOM.Iterable", "ESNext"],
    "allowJs": false,
    "skipLibCheck": true,
    "esModuleInterop": false,
    "allowSyntheticDefaultImports": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "module": "ESNext",
    "moduleResolution": "nodeN",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true,
    "jsx": "react-jsx"
  }
```

**Comment by josip-volarevic** [comment]
Pretty much tried the same ts configuration with nodenext and no luck as well

**Comment by kgilliam125** [comment]
@jordansexton I opened a support ticket w/ Vercel yesterday related to this. They got back and I brought this thread to their attention.

One thing they suggested which doesn't seem to be the issue, but noting in case it helps someone, specifically using Yarn 3 on Vercel:

>We're investigating different options on our side to provide a long term resolution, however in the short term please could you create (or modify if you have it already) your .yarnrc.yml file to include the following setting and then try a few redeploys to see if you experience the same issues on your side:
`cacheFolder: .vercel/cache/yarn`

**Comment by jordansexton** [comment]
Anyone know if there's an eslint rule that will change all imports to explicit .js? Wallet Adapter currently has implicit imports and also index imports so it would be great to have something that will intelligently fix these now, and also scold us if we don't use them.

**Comment by jordansexton** [comment]
@kgilliam125 interesting, thanks for that context. Unfortunately based on the testing in this thread, this also affects projects that don't use yarn, just using npm.

**Comment by jordansexton** [comment]
Woof. Not how I was planning on spending the day!

This is _hopefully_ fixed in https://github.com/solana-labs/wallet-adapter/pull/533 and in the latest alpha release.

Please test it out! Use these dependencies. Probably good to nuke your `node_modules` to be safe.

```json
{
    "@solana/wallet-adapter-base": "^0.9.15-alpha.0",
    "@solana/wallet-adapter-react": "^0.15.17-alpha.0",
    "@solana/wallet-adapter-react-ui": "^0.9.15-alpha.0",
    "@solana/wallet-adapter-wallets": "^0.18.4-alpha.0",
}
```

Unfortunately there is no good automated solution to this. Things that didn't work:

- `NodeNext` [ESM resolution](https://www.typescriptlang.org/docs/handbook/esm-node.html)
- https://github.com/import-js/eslint-plugin-import
- https://github.com/typescript-eslint/typescript-eslint
- https://github.com/weiran-zsd/eslint-plugin-node
- https://github.com/digital-loukoum/tsc-esm
- https://github.com/antongolub/tsc-esm-fix

Things that did work:

- `Node` module resolution
- Regular Expression find/replace
- Lots of trial and error

**Comment by jordansexton** [comment]
Correction: use `"@solana/wallet-adapter-wallets": "^0.18.4-alpha.1"`, the Ledger wallet adapter had an import issue.

**Comment by jordansexton** [comment]
### BOOM

<img width="710" alt="Screen Shot 2022-08-21 at 8 54 40 PM" src="https://user-images.githubusercontent.com/1173161/185823802-d81b7270-581c-4237-8e81-55b041469566.png">

**Comment by aslaker** [comment]
@jordansexton you are a gentleman and a scholar. Thank you so much for putting so much time into figuring this out!

**Comment by jordansexton** [comment]
Production release going out now.

**Comment by jordansexton** [comment]
https://twitter.com/jordaaash/status/1561540587127353344

These should work!

```json
{
    "@solana/wallet-adapter-base": "^0.9.15",
    "@solana/wallet-adapter-react": "^0.15.17",
    "@solana/wallet-adapter-react-ui": "^0.9.15",
    "@solana/wallet-adapter-wallets": "^0.18.4",
}
```

**Comment by Somnus007** [comment]
It works well. Thanks @jordansexton !

**Comment by silvestrevivo** [comment]
Opening a new issue about this change and SvelteKit https://github.com/solana-labs/wallet-adapter/issues/541 since @tiagoappetite reported it. 

It is out of the scope of this project, but since some frameworks like SvelteKit are working with Vite, it would be interesting to find a solution for them.

**Comment by jordansexton** [comment]
Fixed for CJS projects by #542 and published.
